pkg/certificate/manager.go

```diff
@@ -1,43 +1,105 @@
+func NewManager(ctx context.Context, mrcClient MRCClient, serviceCertValidityDuration time.Duration, msgBroker *messaging.Broker) (*Manager, error) {
 	// TODO(#4502): transition this call to a watch function that knows how to handle multiple MRC and can react to changes.
-	mrcs, err := mrcClient.List()
+	mrcEvents, err := mrcClient.Watch(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	client, ca, clientID, err := mrcClient.GetCertIssuerForMRC(mrcs[0])
+	m := &Manager{
+		serviceCertValidityDuration: serviceCertValidityDuration,
+		msgBroker:                   msgBroker,
+	}
+	// We need an MRC in order to get the correct issuers, so handle initial MRC List now
+	mrcs, err := mrcClient.List()
 	if err != nil {
 		return nil, err
 	}
 
-	c := &issuer{Issuer: client, ID: clientID, CertificateAuthority: ca}
+	m.init(mrcs, mrcClient)
+
+	// now that the manager is initialized, start a watch
+	// we wait until the manager is initialized so that there's
+	// no lock contention between the event handler and the initial
+	// state of the manager.
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, open := <-mrcEvents:
+				if !open {
+					// channel was closed; return
+					return
+				}
+
+				m.handleMRCEvent(event)
+			}
+		}
+	}()
 
-	m := &Manager{
-		// The signingIssuer is responsible for signing all newly issued certificates
-		// The validatingIssuer is the issuer that issued existing certificates.
-		// its underlying cert is still in the validating trust store
-		signingIssuer:               c,
-		validatingIssuer:            c,
-		serviceCertValidityDuration: serviceCertValidityDuration,
-		msgBroker:                   msgBroker,
-	}
 	return m, nil
 }
 
+// init initializes the Manager receiver based on the observed state of the cluster at this
+// instant in time. This is why we pass in a slice of MRCs; we know our Watch call will make us initially
+// consistent, we just need a valid state of the world at this instant to kick things off.
+// This method is meant to be called at initialization time, so we don't grab the lock
+func (m *Manager) init(mrcs []*v1alpha2.MeshRootCertificate, mrcClient MRCClient) error {
+	for _, mrc := range mrcs {
+		client, ca, clientID, err := mrcClient.GetCertIssuerForMRC(mrc)
+		if err != nil {
+			return err
+		}
+
+		c := &issuer{Issuer: client, ID: clientID, CertificateAuthority: ca}
+		switch {
+		case mrc.Status.State == "errored":
+			continue
+		case mrc.Status.State == constants.MRCStateComplete && mrc.Status.RotationStage == constants.MRCStageIssuing:
+			m.signingIssuer = c
+			m.validatingIssuer = c
+		case mrc.Status.RotationStage == constants.MRCStageIssuing:
+			m.signingIssuer = c
+		case mrc.Status.RotationStage == constants.MRCStageValidating:
+			m.validatingIssuer = c
+		}
+	}
+
+	if m.signingIssuer == nil || m.validatingIssuer == nil {
+		return ErrNoValidMeshRootCertificates
+	}
+
+	return nil
+}
+
+func (m *Manager) handleMRCEvent(event MRCEvent) {
```

Comment:
based on ordering guarantees, and assuming we have proper validations in the validating webhook, we may want to consider simplifying this, and only accept opaque "updates", (where an update is 1 of an add or an update)

We likely don't need to know the details of prior state (we should have that cached, and as mentioned, old updates may be coalesced), nor do we need to know about deletes, as that should be associated with the promotion of another MRC to active

Reply:
Hm, yeah I see what you mean. I'm good to YAGNI here